docs/source/pytest.rst

```diff
@@ -0,0 +1,489 @@
+
+    import pytest
+    from cocotb_tools.pytest.hdl import HDL
+
+
+    @pytest.fixture(name="sample_module")
+    def sample_module_fixture(hdl: HDL) -> HDL:
+        """Define HDL design by adding HDL source files to it.
+
+        Args:
+            hdl: Fixture created by the cocotb pytest plugin, representing a HDL design.
+
+        Returns:
+            Representation of HDL design with added HDL source files.
+        """
+        hdl.sources = (
+            # List HDL source files,
+            "sample_module.sv",
+        )
+
+        # Build HDL design
+        hdl.build()
+
+        return hdl
+
+
+.. currentmodule:: cocotb_tools
+
+
+:py:deco:`pytest.mark.cocotb`
+==============================
+
+The plugin provides the marker :py:deco:`pytest.mark.cocotb` which allows
+to configure all aspects of cocotb test and cocotb runner.
+
+.. code:: python
+
+    @pytest.mark.cocotb(timescale=("1ns", "1ps"))
+    def test_dut_using_different_timescale(sample_module: HDL) -> None:
+        """Test DUT using different timescale."""
+        sample_module.test()
+
+    @pytest.mark.cocotb(timeout=(200, "ns"))
+    async def test_dut_feature_with_timeout(dut) -> None:
+        """Test DUT feature. It must finish within 200 nanoseconds."""
+
+Additionally, positional arguments of :py:deco:`pytest.mark.cocotb` marker are equivalent to
+``test_module`` argument from :py:meth:`.Runner.test`.
+
+.. code:: python
+
+    @pytest.mark.cocotb("test_dut_tb_1", "test_dut_tb_2")
+    def test_dut_using_different_testbenches(sample_module: HDL) -> None:
+        """Use cocotb tests from ``test_dut_tb_1.py`` and ``test_dut_tb_2.py`` files to test DUT."""
+        sample_module.test()
+
+If no positional arguments were provided to :py:deco:`pytest.mark.cocotb`,
+plugin will load current Python module where :py:deco:`pytest.mark.cocotb` was used as cocotb testbench (Python file with
+cocotb tests).
+
+.. code:: python
+
+    @pytest.mark.cocotb
+    def test_dut_using_default_testbench(sample_module: HDL) -> None:
+        """Test DUT with cocotb tests defined in the same Python file as this test function."""
+        sample_module.test()
+
+    async def test_dut_feature_1(dut) -> None:
+        """Test DUT feature 1."""
+
+If ``toplevel`` argument is empty/non-set, plugin will use name of first test module but without
+``test_*`` prefix or ``*_test`` suffix. For example, if test module was ``test_design`` then
+name of HDL top level design will be ``design``.
+
+.. code:: python
+
+    @pytest.mark.cocotb
+    def test_dut_using_default_toplevel(sample_module: HDL) -> None:
+        """Test DUT with default top level associated with name of test file as this test function."""
+        sample_module.test()
+
+    @pytest.mark.cocotb(toplevel="sample_submodule")
+    def test_dut_using_different_toplevel(sample_module: HDL) -> None:
+        """Test DUT with different top level that was set at fixture level."""
+        sample_module.test()
```

Comment:
I'm not sure about this. In many simulators the build is built for only the given toplevel, so `sample_module` is the build and the toplevel and it shouldn't be reassigned or guessed, etc. This feature should be removed.

Reply:
This is a side-effect of generic for-loop in `HDL` constructor that is assigning marker named arguments to `HDL` attributes that are passed to `Runner`. I will re-factor that and use:

```python
class HDL:
    def __init__(self, request: pytest.FixtureRequest) -> None:
        marker = request.node.get_closest_marker("cocotb_runner")

        self.verbose: bool = marker.kwargs.get("verbose", False)
        ...
```

This will allow avoid that. And I will also update docstring for `@pytest.mark.cocotb_runner` marker + `pytest.rst` documentation.